pip/pip-310.md

```diff
@@ -0,0 +1,121 @@
+# PIP-310 Support custom publish rate limiters
+
+# Background knowledge
+
+Pulsar uses rate limiters to control produce and consume rates. At core, this utility is provided by the RateLimiter.java class.
+There are many places where a rate limiter is used today. Mainly:
+* Produce path:
+  * ResourceGroupPublishLimiter - resource group's produce qps and MBps checks
+  * PrecisePublishLimiter - the topic quota rate limiter in case when `preciseTopicPublishRateLimiterEnable` broker config is `true`
+  * PublishRateLimiterImpl - the topic quota rate limiter in case when `preciseTopicPublishRateLimiterEnable` broker
+  config is `false`
+  * BrokerService - manage total publish rate of the broker
+* Consume path:
+  * DispatchRateLimiter - dispatch rate limiter
+  * BrokerService - manage total dispatch rate of the broker 
+* Misc:
+  * BrokerService - concurrent unloader rate limiter
+  * (Function|Sink|Source)StatsManager
+
+RateLimiter.java has two ways of implementing the rate limit check - precise and poller based. The poller based accumulates allowed permit and check for breach at a regular, configurable, frequency within each second. The precise one checks for breach before handing out permits each time.
```

Comment:
Can you expand about the poller? Each producer holds a set of permits? I don't understand how it works.

Reply:
Adding to the reply to the resource group section - including the details of how poller and precise works - would lead to a very bloated "Background Knowledge" section. I can certainly add both the details, but is there a better approach to this?
Unfortunately, the official documentation about these things is also lacking.

Reply:
Pulsar has a serious knowledge gap in docs, so people reading your PIP have no other place they can can catch up. Sending all your readers to read the code is unacceptable.  I'm not saying add 50 lines, just adding 1 paragraph composed of 3-4 sentences is something easily achievable, and not bloating the design doc at all.